### PR TITLE
Refactor TorchDistribution and wrap torch.distributions.Bernoulli

### PR DESCRIPTION
**Pull request by fritzo**
This adds a working implementation of `pyro.distributions.torch.Bernoulli`. It also simplifies individual torch wrappers by moving the `.batch_shape()` and `.event_shape()` methods up to the parent class `TorchDistribution`.

## Tested

- Ran `make test-torch-dist` against the `test-rsample` branch of PyTorch.

### CODE REVIEW

**Comment by fritzo** [comment]
@neerajprad After this and #647 are merged, I'll address `Categorical`

File: pyro/distributions/torch/bernoulli.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+    def __init__(self, ps=None, logits=None, *args, **kwargs):
+        if (ps is None) == (logits is None):
+            raise ValueError("Got ps={}, logits={}. Either `ps` or `logits` must be specified, "
+                             "but not both.".format(ps, logits))
+        if ps is None:
+            ps = F.sigmoid(logits)
+        eps = get_clamping_buffer(ps)
+        ps = ps.clamp(min=eps, max=1-eps)
+        torch_dist = torch.distributions.Bernoulli(ps)
```
Comment (neerajprad) [inline pyro/distributions/torch/bernoulli.py]: Do any tests fail if we do not do this clamping here?

File: pyro/distributions/torch/gamma.py [hunk 2]
```diff
@@ -13,13 +13,6 @@ class Gamma(TorchDistribution):
...
-    def batch_shape(self, x=None):
-        x_shape = [] if x is None else x.size()
-        shape = torch.Size(broadcast_shape(x_shape, self._param_shape, strict=True))
-        return shape[:-1]
-
-    def event_shape(self):
-        return self._param_shape[-1:]
+        x_shape = torch.Size(broadcast_shape(alpha.size(), beta.size(), strict=True))
+        event_dim = 1
+        super(Gamma, self).__init__(torch_dist, x_shape, event_dim, *args, **kwargs)
```
Comment (neerajprad) [inline pyro/distributions/torch/gamma.py]: It's good to see all of this getting absorbed into the wrapper class!

File: pyro/distributions/torch/bernoulli.py (hunk 1 above)
Comment (fritzo) [reply pyro/distributions/torch/bernoulli.py]: Yes, your `_flow` tests fail. I'm glad you implemented them.

**Comment by fritzo** [comment]
Thanks for reviewing!

File: pyro/distributions/torch/bernoulli.py (hunk 1 above)
Comment (neerajprad) [reply pyro/distributions/torch/bernoulli.py]: Ahh..I see. Makes sense! Will add logit support in PyTorch soon.

**Pull request merged**